lxd/instance.go

```diff
@@ -83,6 +83,40 @@ func instanceImageTransfer(s *state.State, r *http.Request, projectName string,
 	return nil
 }
 
+func ensureImageIsLocallyAvailable(s *state.State, r *http.Request, img *api.Image, project string, instanceType instancetype.Type) error {
+	// Check if the image is available locally or it's on another member.
+	// Ensure we are the only ones operating on this image. Otherwise another instance created at the same
+	// time may also arrive at the conclusion that the image doesn't exist on this cluster member and then
+	// think it needs to download the image and store the record in the database as well, which will lead to
+	// duplicate record errors.
+	unlock := imageOperationLock(img.Fingerprint)
+
+	memberAddress, err := s.DB.Cluster.LocateImage(img.Fingerprint)
+	if err != nil {
+		unlock()
+		return fmt.Errorf("Failed locating image %q: %w", img.Fingerprint, err)
+	}
+
+	if memberAddress != "" {
+		// The image is available from another node, let's try to import it.
+		err = instanceImageTransfer(s, r, project, img.Fingerprint, memberAddress)
+		if err != nil {
+			unlock()
+			return fmt.Errorf("Failed transferring image %q from %q: %w", img.Fingerprint, memberAddress, err)
+		}
+
+		// As the image record already exists in the project, just add the node ID to the image.
+		err = s.DB.Cluster.AddImageToLocalNode(project, img.Fingerprint)
+		if err != nil {
+			unlock()
+			return fmt.Errorf("Failed adding transferred image %q record to local cluster member: %w", img.Fingerprint, err)
+		}
+	}
+
+	unlock() // Image is available locally.
```

Comment:
Now this is in its own function we can use a defer to unlock as soon as the lock is taken.